cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -85,67 +87,114 @@ message Result {
 }
 ```
 
-Method : GetEndpointMetadataByIP
-
-Request : 
-```
-message FQDNSelector {
-  string ip = 1;
-}
-```
-Response : 
-```
-message Result {
-    uint32 endpoint_id = 1;
-    uint32 identity = 2;
-}
-```
-
 ### Tracking policy updates to SDP instances
 
 Since SDP can be deployed out of band and users can choose to completely disable built-in proxy to run multiple instances of SDP, agent should be prepared to handle multiple instances. In order to ensure all instances  have upto date policy revisions, agent will maintain a mapping of ACKed policy revision numbers against stream ID.
-
+Since policy revision numbers from Cilium are reset when the local Cilium instance restarts, we need to unconditionally send the policy updates to the SDP instance on agent restart.
 
 ## Impacts / Key Questions
 
 
 ### Discovering endpoint metadata from SDP
 
-In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. Since the actual source of truth are the bpf maps. Should the SDP be allowed to directly read from bpf maps ?
+In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. Since the actual source of truth are the bpf maps. And SDP is expected to resolve the DNS queries even when agent is down, it is better to read the information directly from the bpf maps. We choose the bpf option as it is more reliable and does not rely on agent availability. It also aligns with having a HA DNS proxy.
 
+#### Option 1a: gRPC method
 
-### Option 1: gRPC method
+Get the endpoint ID and identity for a given IP address by making a gRPC call to cilium agent.
 
-#### Pros
+##### Pros
 
 * Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
 * All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
 
-#### Cons
+##### Cons
 
 * Reliance on gRPC call in the hot-path
+* Does not cache the data for endpoints that never had a DNS query.
 * In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
 
-### Option 2: Read from bpf maps
+#### Option 1b: Listen to ipcache updates via grpc
 
-#### Pros
+Similar to envoy, SDP can listen to ipcache updates and maintain a local cache of IP to endpoint ID mappings via grpc. This changes the way DNS proxy currently behaves i.e pull based model to push based model.
```

Comment:
Does the L7 proxy list the entire ipcache on startup already ?

Reply:
I am not sure how they on the startup but didn't find anything that talks about the reading the bpf maps directly. They listen on ipcache updates as far as I can tell.

Reply:
Added an open ended question about if envoy is doing the identity resolution in the section: ` Q: Why did we choose to read from bpf maps ?`